### PR TITLE
[AutoDiff] Derive `Differentiable.zeroTangentVectorInitializer`.

### PR DESCRIPTION
**Pull request by dan-zheng**
`Differentiable` conformance derivation now supports
`Differentiable.zeroTangentVectorInitializer`.

`zeroTangentVectorInitializer` is a closure that produces a zero tangent vector,
capturing minimal necessary information from `self`.

It is an instance property, unlike the static property `AdditiveArithmetic.zero`,
and should be used by the differentiation transform for correctness.

Remove `zeroTangentVectorInitializer` dummy default implementation.
Primitive `Differentiable`-conforming types now must implement it.

Update stdlib `Differentiable` conformances and tests.

Resolves TF-1007.
Progress towards TF-1008: differentiation correctness for projection operations.

---

Examples:

```swift
import _Differentiation

struct Struct<T: Differentiable, U: Differentiable>: Differentiable {
  var x: T
  var y: U

  // Compiler synthesizes:
  // var zeroTangentVectorInitializer: () -> TangentVector {
  //   { [xZeroTanInit = x.zeroTangentVectorInitializer,
  //      yZeroTanInit = y.zeroTangentVectorInitializer] in
  //     return TangentVector(x: xZeroTanInit(), y: yZeroTanInit())
  //   }
  // }
}

let s = Struct(x: [1, 2, 3], y: [[4, 5, 6], [], [2]])
// `Differentiable.zeroTangentVector` default implementation calls
// `zeroTangentVectorInitializer`.
print(s.zeroTangentVector)

// Before (via dummy `zeroTangentVectorInitializer` default implementation):
// TangentVector(x: [], y: [])

// After (via derived conformances):
// TangentVector(x: [0.0, 0.0, 0.0], y: [[0.0, 0.0, 0.0], [], [0.0]])
```

```swift
struct CustomTangentVector<T: Differentiable, U: Differentiable>: Differentiable {
  var x: T
  var y: U

  typealias TangentVector = T.TangentVector
  mutating func move(along direction: TangentVector) {}

  // Memberwise synthesis is not possible for custom `TangentVector` type.
  // Fallback synthesis is done instead:
  // var zeroTangentVectorInitializer: () -> TangentVector {
  //   { TangentVector.zero }
  // }
}
```

### CODE REVIEW

File: lib/Sema/MiscDiagnostics.cpp [hunk 1]
```diff
@@ -1449,6 +1449,16 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
...
+      // If the closure is in an implicit declaration, then it doesn't need
+      // qualification. Diagnostics would show an invalid source location.
+      //
+      // Currently used for `Differentiable.zeroTangentVectorInitializer`
+      // derived conformances.
+      if (CE->isImplicit()) {
+        auto *decl = CE->getInnermostDeclarationDeclContext();
+        if (decl && decl->isImplicit())
+          return false;
+      }
```
Comment (dan-zheng) [inline lib/Sema/MiscDiagnostics.cpp:1461]: Note: this special case is necessary to avoid diagnosing implicit `self` `DeclRefExpr` qualification in synthesized `zeroTangentVectorInitializer` declarations.

File: lib/Sema/DebuggerTestingTransform.cpp [hunk 2]
```diff
@@ -33,35 +34,6 @@ using namespace swift;
 
 namespace {
 
-/// Find available closure discriminators.
-///
-/// The parser typically takes care of assigning unique discriminators to
-/// closures, but the parser is unavailable to this transform.
-class DiscriminatorFinder : public ASTWalker {
```
Comment (dan-zheng) [inline lib/Sema/DebuggerTestingTransform.cpp:40]: Note: I moved `DiscriminatorFinder` to a shared location `lib/Sema/CodeSynthesis.h`.

**Comment by dan-zheng** [comment]
@CodaFi: I'll add you as reviewer in case you'd like to look at any of these derived conformances changes. `lib/Sema/DerivedConformanceDifferentiable.cpp` has some room for cleanup that I'd like to investigate later.

**Comment by dan-zheng** [comment]
@swift-ci Please test

File: test/AutoDiff/Sema/DerivedConformances/class_differentiable.swift [hunk 3]
```diff
@@ -109,28 +109,13 @@ final class VectorSpacesEqualSelf: Differentiable & DummyAdditiveArithmetic {
   var w: Float
   var b: Float
   typealias TangentVector = VectorSpacesEqualSelf
+  var zeroTangentVectorInitializer: () -> TangentVector { fatalError() }
```
Comment (rxwei) [inline test/AutoDiff/Sema/DerivedConformances/class_differentiable.swift:112]: Shouldn't these use the default implementation (`.zero`)?  Is it because derivation won't be triggered if you provide a default implementation?

Comment (dan-zheng) [reply test/AutoDiff/Sema/DerivedConformances/class_differentiable.swift:112]: > Is it because derivation won't be triggered if you provide a default implementation?

Yes, we have to remove the dummy default implementation of `Differentiable.zeroTangentVectorInitializer` using `TangentVector.zero` because its existence prevents derived conformances from triggering.

This requires "primitive" `Differentiable`-conforming types to implement `zeroTangentVectorInitializer`.

File: include/swift/AST/KnownIdentifiers.def [hunk 4]
```diff
@@ -223,6 +223,8 @@ IDENTIFIER(move)
 IDENTIFIER(pullback)
 IDENTIFIER(TangentVector)
 IDENTIFIER(zero)
+IDENTIFIER(zeroTangentVector)
```
Comment (rxwei) [inline include/swift/AST/KnownIdentifiers.def:226]: Why is this needed?

Comment (dan-zheng) [reply include/swift/AST/KnownIdentifiers.def:226]: Thanks, this isn't needed. Removed.

**Review by rxwei — COMMENTED** [review]
Can you derive a default implementation for `zeroTangentVectorInitializer` even when the user has a custom tangent vector? If you do that then this PR won't be source-breaking.

**Comment by dan-zheng** [comment]
> Can you derive a default implementation for `zeroTangentVectorInitializer` even when the user has a custom tangent vector? If you do that then this PR won't be source-breaking.

I asked offline for clarification. What @rxwei has in mind is: deriving `zeroTangentVectorInitializer` as `{ TangentVector.zero }` for types defining a custom `TangentVector`.

@marcrasi: do you agree whether this makes sense and is desirable? I think you have experience with custom `TangentVector` types from [borglab/SwiftFusion](https://github.com/borglab/SwiftFusion).

This behavior should be possible but requires more work. I'd like to reach consensus before exploring.

**Comment by marcrasi** [comment]
I have some thoughts about default synthesis of zeroTangentVectorInitializer.

**The ultimate ideal behavior**

* We clarify the semantics of `AdditiveArithmetic` to require that `static var zero` always "makes sense". (Depending on how you interpret the documentation, this is already how the semantics are, and it is wrong for `Tensor` to conform to `AdditiveArithmetic`).
* Therefore we remove the conformances of `Array`, `Tensor`, etc to `AdditiveArithmetic`.
* We introduce a new `protocol X` that captures the concept of an additive arithmetic without zero. `AdditiveArithmetic` refines `X`.
* We relax the `Differentiable` requirement that `TangentVector: AdditiveArithmetic`. Make it `TangentVector: X` instead.
* When `Differentiable` synthesis sees a custom `TangentVector` that conforms to `AdditiveArithmetic`, it synthesizes a `zeroTangentVectorInitializer`. This is completely correct and safe because of the semantics of `AdditiveArithmetic`.
* When `Differentiable` synthesis sees a custom `TangentVector` that does not conform to `AdditiveArithmetic`, it is unable to synthesize `zeroTangentVectorInitializer`.

**What to do today**

Yes, derive `zeroTangentVectorInitializer` as `{ TangentVector.zero }` for types defining a custom `TangentVector`. This is consistent with the ultimate ideal behavior, and therefore a step in the right direction.

**Comment by marcrasi** [comment]
**From a usability point of view**

It's already pretty difficult to define a custom `TangentVector`. If we don't automatically synthesize `zeroTangentVectorInitialzer` for custom `TangentVector`s, then it becomes even more difficult. This is a step in the wrong direction for usability.

Therefore the usability point of view also says that, it's good to derive `zeroTangentVectorInitializer` as `{ TangentVector.zero }` for types defining a custom `TangentVector`.

File: lib/Sema/MiscDiagnostics.cpp (hunk 1 above)
Comment (marcrasi) [reply lib/Sema/MiscDiagnostics.cpp:1461]: Is it possible to synthesize `self.xxx` instead of adding this special case here?

File: stdlib/public/Differentiation/ArrayDifferentiation.swift [hunk 5]
```diff
@@ -168,9 +172,7 @@ extension Array: Differentiable where Element: Differentiable {
   /// A closure that produces a `TangentVector` of zeros with the same
   /// `count` as `self`.
   public var zeroTangentVectorInitializer: () -> TangentVector {
-    { [count = self.count] in
-      TangentVector(.init(repeating: .zero, count: count))
-    }
+    { TangentVector(map { $0.zeroTangentVector }) }
```
Comment (marcrasi) [inline stdlib/public/Differentiation/ArrayDifferentiation.swift:175]: This is interesting, good catch.

However, this is capturing `self`. It should instead capture all the elements' `zeroTangentVectorInitializer`s.

This sounds potentially terrible for performance for a few reasons:

* If the closure captures anything, then it probably does one heap allocation per closure. (Maybe the compiler is clever enough to realize that all the closure captures are the same and only do one heap allocation for all of them? That would be very magical.
* In the case where `TangentVector(.init(repeating: .zero, count: count))` is correct, we would really like to use `count` so that we can avoid iterating over the array.

I can think of one solution that involves defining a few different differentiable views of array that are specialized to work efficiently in different cases. This solution can be implemented in client code, so we could leave the stdlib as-is for now and eventually move the solution into the stdlib if we discover that it really is necessary.

I wonder if there are any other solutions that are easier for users to use?

File: lib/Sema/DerivedConformanceDifferentiable.cpp [hunk 6]
```diff
@@ -281,6 +299,139 @@ deriveBodyDifferentiable_move(AbstractFunctionDecl *funcDecl, void *) {
...
+  auto *captureList = CaptureListExpr::create(C, {captureEntry}, closure);
+  captureList->setImplicit();
+
+  ASTNode returnStmt = new (C) ReturnStmt(SourceLoc(), captureList, true);
+  auto *braceStmt =
+      BraceStmt::create(C, SourceLoc(), returnStmt, SourceLoc(), true);
+  return std::pair<BraceStmt *, bool>(braceStmt, false);
+}
+
+/// Synthesize body for `move(along:)`.
```
Comment (marcrasi) [inline lib/Sema/DerivedConformanceDifferentiable.cpp:426]: This comment is wrong?

File: lib/Sema/DerivedConformanceDifferentiable.cpp [hunk 7]
```diff
@@ -281,6 +299,139 @@ deriveBodyDifferentiable_move(AbstractFunctionDecl *funcDecl, void *) {
   return deriveBodyDifferentiable_method(funcDecl, C.Id_move, C.Id_along);
 }
 
+/// Synthesize body for a `Differentiable` method requirement.
+static std::pair<BraceStmt *, bool>
+deriveBodyDifferentiable_zeroTangentVectorInitializer_getter(
```
Comment (marcrasi) [inline lib/Sema/DerivedConformanceDifferentiable.cpp:304]: It would be useful to have comments inline with pseudocode corresponding to the code that is being generated.

There is some pseudocode but it's confusingly unrelated to the actual code that is being generated. Like the "`x.move(along: direction.x)`". I don't think it's synthesizing any `move(along:)` calls here.

**Comment by rxwei** [comment]
The root problem is that types like `Tensor` have dynamic ranks, so the tangent vector type of a `Tensor` is in fact value-dependent.  I wish Swift supports things like path-dependent types one day.

```swift
@derivative(of: foo)
func _(x: T) -> (value: U, differential: @differentiable(linear) (x.T) -> value.TangentVector
```

However, `zeroTangentVectorInitializer()` will be a required workaround in the foreseeable future...

File: stdlib/public/Differentiation/ArrayDifferentiation.swift (hunk 5 above)
Comment (rxwei) [reply stdlib/public/Differentiation/ArrayDifferentiation.swift:175]: Yeah, calling `.zeroTangentVector` defeats the whole purpose of `zeroTangentVectorInitializer`. How about:
```swift
{ [zeroInits = map(\.zeroTangentVectorInitializer)] in
    TangentVector(zeroInits.map { $0() })
}
```

Comment (dan-zheng) [reply stdlib/public/Differentiation/ArrayDifferentiation.swift]: Thanks for the catch!

I applied @rxwei's suggestion: `Array.zeroTangentVectorInitializer` captures only elements' `zeroTangentVectorInitializer` closures.

**Review by rxwei — APPROVED** [review]

File: lib/Sema/MiscDiagnostics.cpp (hunk 1 above)
Comment (dan-zheng) [reply lib/Sema/MiscDiagnostics.cpp]: > Is it possible to synthesize `self.xxx` instead of adding this special case here?

`self.xxx` was actually synthesized as an implicit `DeclRefExpr`. However, there's no distinction between an implicit `self` `DeclRefExpr` generated via derived conformances or via other parts of type-checking, which made this special case necessary.

Comment (dan-zheng) [reply lib/Sema/MiscDiagnostics.cpp]: I was able to revert this special case!

The previously synthesized ASTs were actually wrong. Synthesized `Differentiable.zeroTangentVectorInitializer` closures should never capture `self` (i.e. contain implicit `self` `DeclRefExpr`s).

```swift
import _Differentiation
struct Struct<T: Differentiable, U: Differentiable>: Differentiable {
  var x: T
  var y: U

  // Previously synthesized, incorrectly captures `self`:
  // var zeroTangentVectorInitializer: () -> TangentVector {
  //   {
  //     // Accesses to `self.x` and `self.y` cause `self` to be captured.
  //     TangentVector(x: x.zeroTangentVectorInitializer(),
  //                   y: y.zeroTangentVectorInitializer())
  //   }
  // }

  // Now synthesized, correct:
  // var zeroTangentVectorInitializer: () -> TangentVector {
  //   { [xZeroTanInit = x.zeroTangentVectorInitializer,
  //      yZeroTanInit = y.zeroTangentVectorInitializer] in
  //     TangentVector(x: xZeroTanInit(), y: yZeroTanInit())
  //   }
  // }
}
```

File: lib/Sema/DerivedConformanceDifferentiable.cpp (hunk 6 above)
Comment (dan-zheng) [reply lib/Sema/DerivedConformanceDifferentiable.cpp]: Fixed, thanks!

File: lib/Sema/DerivedConformanceDifferentiable.cpp (hunk 7 above)
Comment (dan-zheng) [reply lib/Sema/DerivedConformanceDifferentiable.cpp]: Thanks for the catch! I copied the code for synthesized `move(along:)` and forgot to update comments.

Comments are fixed now.

**Comment by dan-zheng** [comment]
Thanks for the feedback above!

> Yes, derive `zeroTangentVectorInitializer` as `{ TangentVector.zero }` for types defining a custom `TangentVector`. This is consistent with the ultimate ideal behavior, and therefore a step in the right direction.

I implemented this behavior. There are now two cases for `Differentiable.zeroTangentVectorInitializer` derivation:
1. Memberwise derivation, when `TangentVector` type can be initialized memberwise.
2. Fallback `{ TangentVector.zero }` derivation.

The result is improved usability: almost no types (including "primitive" `Differentiable`-conforming ones) are _required_ to define `Differentiable.zeroTangentVectorInitializer`.

---

`Differentiable.zeroTangentVectorInitializer` derivation is now also fixed so that synthesized closures never capture `self`.

<details>

```swift
import _Differentiation
struct Struct<T: Differentiable, U: Differentiable>: Differentiable {
  var x: T
  var y: U

  // Previously synthesized, incorrectly captures `self`:
  // var zeroTangentVectorInitializer: () -> TangentVector {
  //   {
  //     // Accesses to `self.x` and `self.y` cause `self` to be captured.
  //     TangentVector(x: x.zeroTangentVectorInitializer(),
  //                   y: y.zeroTangentVectorInitializer())
  //   }
  // }

  // Now synthesized, correct:
  // var zeroTangentVectorInitializer: () -> TangentVector {
  //   { [xZeroTanInit = x.zeroTangentVectorInitializer,
  //      yZeroTanInit = y.zeroTangentVectorInitializer] in
  //     TangentVector(x: xZeroTanInit(), y: yZeroTanInit())
  //   }
  // }
}
```

</details>

Ready for re-review.

**Comment by dan-zheng** [comment]
@swift-ci Please smoke test

File: test/AutoDiff/Sema/DerivedConformances/derived_zero_tangent_vector_initializer.swift [hunk 8]
```diff
@@ -0,0 +1,106 @@
+// RUN: %target-swift-frontend -emit-silgen %s | %FileCheck %s
+
+// Check `Differentiable.zeroTangentVectorInitializer` derivation.
+// There are two case:
+// 1. Memberwise derivation.
+// 2. `{ TangentVector.zero }` fallback derivation.
+
+import _Differentiation
+
+struct MemberwiseTangentVector: Differentiable {
```
Comment (rxwei) [inline test/AutoDiff/Sema/DerivedConformances/derived_zero_tangent_vector_initializer.swift:10]: This PR seems to be missing `class` tests.

Comment (dan-zheng) [reply test/AutoDiff/Sema/DerivedConformances/derived_zero_tangent_vector_initializer.swift]: Added `zeroTangentVectorInitializer` SILGen tests for classes and an enum.

**Review by rxwei — APPROVED** [review]

**Comment by dan-zheng** [comment]
@swift-ci Please smoke test

**Comment by dan-zheng** [comment]
@swift-ci Please clean smoke test Linux
